rem_vertex! bugfix

jarbus opened the pull request:
Resolves an issue where properties that point from the last node to the last node don't get updated with the new vertex number, referenced in #87

The current version of MetaGraphs.jl fails with the following example:
```julia
using LightGraphs, MetaGraphs
g = MetaDiGraph(2)
add_edge!(g, 2, 2)
rem_vertex!(g, 1)
```
Yields

```
ERROR: LoadError: KeyError: key 1 not found
Stacktrace:
 [1] getindex
   @ ./dict.jl:482 [inlined]
 [2] rem_vertex!(g::MetaDiGraph{Int64, Float64}, v::Int64)
   @ MetaGraphs ~/.julia/packages/MetaGraphs/UlMik/src/MetaGraphs.jl:184
```

kescobo commented:
Glorious!

sbromberger commented:
@kescobo - now with induced subgraphs. Check this out:
```
filt = filter_vertices!(g, :color, "red") # select all the red nodes
g2 = g[filt] # give me a graph of just the red nodes and the links between them
```

kescobo commented:
@sbromberger Cool! But I'm a bit confused - is `filter_vertices!()` actually modifying `g`? If yes, why do you need the second step. And if not, why the `!` in the function signature?

kescobo commented:
Might be cool to have a `Range`-like object that's returned from a call to `filter_X`, that can be used to index or make a view with the syntax in line 2 there.

jpfairbanks commented:
This is awesome!

Maybe we need domain filtering too. So you can do filtering in the vertex domain rather than the property domain.

sbromberger commented:
Now with generic functions:
```
f = filter_edges(mg, (g, e) -> has_prop(g, e, :weight) && get_prop(g, e, :weight) > 0.3)
```

sbromberger commented:
>  is filter_vertices!() actually modifying g? If yes, why do you need the second step. And if not, why the ! in the function signature?

Oops. That was a typo. Non-mutating.

jpfairbanks commented:
Yeah `filter_vertices` creates an indexing object and then the `getindex` creates the subgraph.

sbromberger commented:
@kescobo 

> Might be cool to have a Range-like object that's returned from a call to filter_X, that can be used to index or make a view with the syntax in line 2 there.

Calling `induced_subgraph` directly gives you the map (it returns a vector of "new vertex ID" to "old vertex ID").

jpfairbanks commented on src/MetaGraphs.jl:
```diff
@@ -31,7 +32,12 @@ export
     clear_props!,
-    set_weightfield!
+    set_weightfield!,
+    set_defaultweight!,
```
I am a little worried about this JAVA style interface, why do you need getters and setters? They just get and set the actual fields.

what about using the names `defaultweight(g)` for the getter and `defaultweight!(g, value)` for the setter?
You still get the encapsulation you want by having the function call in there, but you don't make it read like ☕️ ☕️ ☕️ ☕️ ☕️ ☕️ ☕️

sbromberger replied:
I am trying to make explicit when it's OK to access an attribute in a struct. I agree it sucks. I like your idea, though, and I'm going to do that throughout now.

sbromberger replied:
and I hate java. :)

sbromberger commented:
> Maybe we need domain filtering too. So you can do filtering in the vertex domain rather than the property domain.

We have that now, right?

jpfairbanks commented:
But does it respect the metadata? When you renumber the vertices (because you removed some), then you have to renumber the edge metadata dictionaries.

Unless you just leave holes for what you deleted.

sbromberger commented:
> But does it respect the metadata? When you renumber the vertices (because you removed some), then you have to renumber the edge metadata dictionaries.

Yes. Check out `_copy_props!()`.

jpfairbanks commented:
Oh yeah that does what I was thinking.

sbromberger commented:
I'm going to go ahead and merge this - we can fix things up later.

jarbus committed:
rem_vertex! bugfix

Resolves an issue where properties that point from the last node to the last node don't get updated with the new vertex number

jarbus committed:
adding test for #87

jarbus commented:
@matbesancon would it be possible to merge this in

matbesancon commented:
thanks!
